### ISSUE TITLE
Deleting ndarray.real on an instance segfaults (migrated from Trac #2046)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3593]
Original ticket http://projects.scipy.org/numpy/ticket/2046
Reported 2012-02-10 by atmention:alex, assigned to atmention:charris.

Tested on the latest numpy master from github:

```
(e268cfc7-ff3f-4d97-a02c-27481c8a8186)alexatmention:alex-gaynor-laptop:~/.virtualenvs/e268cfc7-ff3f-4d97-a02c-27481c8a8186$ python
Python 2.7.2+ (default, Oct  4 2011, 20:06:09) 
[GCC 4.6.1] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
>>> del numpy.arange(5).real
Segmentation fault
```

**Comment by thouis** [comment]
Comment in Trac by atmention:alex, 2012-02-10

Deleting `ndarray.imag` fails in a similar manner, but only on complex arrays, other arrays raise an error about not having an imaginary part.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-02-13

This seems to be a problem with getter functions returning writable objects:

```
In [5]: del ones(2).base
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/home/charris/<ipython-input-5-f9c0ca6fc60a> in <module>()
----> 1 del ones(2).base

AttributeError: attribute 'base' of 'numpy.ndarray' objects is not writable

In [6]: del ones(2).dtype
Segmentation fault (core dumped)
```

I'm guessing this is a reference counting problem.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-02-13

Oops, sorry I overwrote the original.

Note that if a getter isn't used we get

```
In [3]: del ones(3)
  File "<ipython-input-3-b548f34f731a>", line 1
SyntaxError: can't delete function call (<ipython-input-3-b548f34f731a>, line 1)


In [4]: del ones(3).view(double)
  File "<ipython-input-4-6d0dbe42344e>", line 1
SyntaxError: can't delete function call (<ipython-input-4-6d0dbe42344e>, line 1)
```

which is a strange message. But it looks like this might almost be considered a Python bug where python del can't recognize getters/setters.

**Comment by thouis** [comment]
Comment in Trac by atmention:alex, 2012-02-13

That's not a strange message, there's no protocol for Python to invoke if you try to delete a function call, so it raises a SyntaxError at parse time.

For Python code, when del is used on an attribute it invokes **delattr** (or **delete** on a descriptor), for C code with a descriptor it'll invoke the setter with a NULL argument.  My guess is numpy isn't handling NULL arguments correctly.

This isn't a bug in Python.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-02-13

It's not a standard attribute, it's a getter function call underneath the hood. I don't know if Python can recogize that, however. But what I think the writeable error message means is that the corresponding (setter) function is called with a NULL value (meaning delete) and we aren't checking that.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-02-13

Yep, that is what is was. Now I can do

```
In [1]: del ones(3).dtype
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/home/charris/<ipython-input-1-92786cf759d7> in <module>()
----> 1 del ones(3).dtype

AttributeError: Can't delete array dtype
```

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-02-13

Duplicate 0f 2044, should be closed when this one is.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-02-13

Duplicate of #2044, fixed in 7e202a2.
